File: 2.download_images.py
#!/usr/bin/python3
#
# Download images from EcoTaxa
#
# (c) 2022 Jean-Olivier Irisson, GNU General Public License v3

import os
import yaml

import shutil
import urllib.request

from tqdm import tqdm
from pyarrow.parquet import read_table

# read config
with open(r'config.yaml') as config_file:
    cfg = yaml.safe_load(config_file)
print('### Download images for {}'.format(cfg['dataset']))

# prepare storage
data_dir = os.path.join(os.path.expanduser(cfg['base_dir']), cfg['dataset'])
img_dir = os.path.join(data_dir, 'orig_imgs')
os.makedirs(img_dir, exist_ok=True)

# read data from EcoTaxa
df = read_table(os.path.join(data_dir, 'orig_extraction.parquet')).to_pandas()

# detect image extension
file,ext = os.path.splitext(df['img.file_name'][0])
# TODO extract file extension for each file, like in step 4.
# name image according to internal object_id to ensure uniqueness
df['dest_path'] = [os.path.join(img_dir, str(this_id)+ext) for this_id in df['objid']]

# download images (with a nice progress bar)
vault_path = '/remote/ecotaxa/vault'
for i in tqdm(range(df.shape[0])):
  # if the file has not been copied already
  if not os.path.isfile(df['dest_path'][i]):
    # copy from vault
    if os.path.isdir(vault_path):
      res = shutil.copyfile(
        src=os.path.join(vault_path, df['img.file_name'][i]),
        dst=df['dest_path'][i]
      )
    # or copy through the internet
    else:
      res = urllib.request.urlretrieve(
        url='https://ecotaxa.obs-vlfr.fr/vault/'+df['img.file_name'][i],
        filename=df['dest_path'][i]
      )

n_imgs = len(os.listdir(img_dir))
print('  {} images in {}'.format(n_imgs, img_dir))

File: 2.download_images.R
#
## Download images from EcoTaxa
#
#
#load packages
library("ecotaxarapi")
library("tools")
library('utils')
library("stringr")
library("yaml")
library("arrow")
library("progress")
library("dplyr")
library("future.apply")
library("parallel")



# Read config file
cfg <- read_yaml("config.yaml")
print(paste0('### Download images for ', cfg$dataset))


# Prepare storage for orig_imgs
data_dir <- file.path((cfg$base_dir), cfg$dataset)
img_dir <- file.path(data_dir, 'orig_imgs')
dir.create(img_dir, recursive = TRUE)

# Read data from EcoTaxa
df <- read_parquet(file.path(data_dir, 'orig_extraction.parquet'))

# Detect image extension
ext <- file_ext(df$img.file_name)[1]

# Name image according to internal object_id to ensure uniqueness
df$dest_path = file.path(img_dir, paste0(df$objid, '.', ext))

# Check files existence 
file_exists_list <- mclapply(df$dest_path, file.exists, mc.cores = detectCores()/4)
df$file_exists <- file_exists_list

# Filter missing files
df_missing <- df %>% filter(file_exists == FALSE)

# Check vault path existence
vault_path <- '/remote/ecotaxa/vault'
vault <- file.exists(vault_path)

# Function to download/copy files
process_file <- function(row_i, dfmissing, vault) {
  system(paste("echo 'now processing row:",row_i,"' on ", nrow(dfmissing), "row"))
  if (vault) {
    file.copy(
      from = file.path(vault_path, dfmissing$img.file_name[row_i]),
      to = dfmissing$dest_path[row_i]
    )
  } else {
    download.file(
      url = paste0('https://ecotaxa.obs-vlfr.fr/vault/', dfmissing$img.file_name[row_i]),
      destfile = dfmissing$dest_path[row_i],
      quiet = TRUE
    )
  }
}

# Apply fonction in parallel
print("Copying or downloading images, be patient")
mclapply(1:nrow(df_missing), process_file, dfmissing = df_missing, vault = vault, mc.cores = detectCores()/4)

# Check images in directory
n_imgs <- length(list.files(img_dir))
print(paste0(n_imgs, ' images in ', img_dir))

